### PR TITLE
Preserve custom type of auto-generated id property

### PR DESCRIPTION
**Pull request by frbuceta**
Implements #3602 

## Checklist

👉 [Read and sign the CLA (Contributor License Agreement)](https://cla.strongloop.com/agreements/strongloop/loopback-next) 👈

- [x] `npm test` passes on your machine
- [x] New tests added or **existing tests modified to cover all changes**
- [x] Code conforms with the [style guide](http://loopback.io/doc/en/contrib/style-guide.html)
- API Documentation in code was updated
- Documentation in [/docs/site](../tree/master/docs/site) was updated
- Affected artifact templates in `packages/cli` were updated
- Affected example projects in `examples/*` were updated

👉 [Check out how to submit a PR](https://loopback.io/doc/en/lb4/submitting_a_pr.html) 👈

### CODE REVIEW

**Comment by frbuceta** [comment]
Can someone help me with the test? It seems not to be happening correctly

**Comment by derdeka** [comment]
> Can someone help me with the test? It seems not to be happening correctly

The changes looks reasonable to me and local unit test works successfully.

TASK=test-repository-mongodb and TASK=test-repository-mysql fail in travis ci.
The `useDefaultIdType=false` seems not to be used by the juggler.
For mysql the following ddl is still created:
```
CREATE TABLE `User` (
`id` TEXT NOT NULL AUTO_INCREMENT PRIMARY KEY,
`name` VARCHAR(512) NULL,
`roles` TEXT NULL,
`address` TEXT NULL
)
```
Type `TEXT` ad `AUTO_INCREMENT` are not a valid combination.

**Comment by frbuceta** [comment]
> > Can someone help me with the test? It seems not to be happening correctly
> 
> The changes looks reasonable to me and local unit test works successfully.
> 
> TASK=test-repository-mongodb and TASK=test-repository-mysql fail in travis ci.
> The `useDefaultIdType=false` seems not to be used by the juggler.
> For mysql the following ddl is still created:
> 
> ```
> CREATE TABLE `User` (
> `id` TEXT NOT NULL AUTO_INCREMENT PRIMARY KEY,
> `name` VARCHAR(512) NULL,
> `roles` TEXT NULL,
> `address` TEXT NULL
> )
> ```
> 
> Type `TEXT` ad `AUTO_INCREMENT` are not a valid combination.

I certainly did not know that I took code from outside the repository to test, thanks for your help.
As always here I am constantly learning

File: packages/repository/src/model.ts [hunk 1]
```diff
@@ -106,6 +106,11 @@ export class ModelDefinition {
     const definition = (definitionOrType as PropertyDefinition).type
       ? (definitionOrType as PropertyDefinition)
       : {type: definitionOrType};
+
+    if (definition.id === true && definition.generated === true) {
```
Comment (bajtos) [inline packages/repository/src/model.ts]: To be safe, I am proposing to extend this check to apply `useDefaultIdType: false` only when an explicit `type` is provided.

This property should set `useDefaultIdType: false`:

```ts
@property({
  type: 'string',
  id: true,
  generated: true
})
id: string;
```

This property should NOT set `useDefaultIdType: false`:

```ts
@property({
  id: true,
  generated: true
})
id: string | number;
```

Thoughts?

**Comment by bajtos** [comment]
Thank you @frbuceta for the pull request, this is great!

> Type `TEXT` ad `AUTO_INCREMENT` are not a valid combination.

Yeah, this is a problem we have recently discovered too, see https://github.com/strongloop/loopback-next/issues/2398. I think we need to fix our connectors to correctly support auto-generated string ids (primary keys) before we can move forward with this pull request.

@agnes512 you were looking in to #2398, what do you think?

**Comment by bajtos** [comment]
> TASK=test-repository-mongodb and TASK=test-repository-mysql fail in travis ci.
The `useDefaultIdType=false` seems not to be used by the juggler.

On the second thought, I think this is a bug in the way how our shared test suite is creating models. 

The way how this test suite is supposed to work, is that the connector specifies what type is used for primary keys (number vs. string) and the test suite honors that type when defining new models.

I suspect the code building `User` model is not taking into account database-specific type and always uses `string`.

AFAICT, `User.id` property is defined here:

https://github.com/strongloop/loopback-next/blob/4995057ad4c938d6751c329632eb44130ac1567a/packages/repository-tests/src/crud/nested-model-properties.suite.ts#L91-L95

Here is the correct definition of the PK property in a different model:

https://github.com/strongloop/loopback-next/blob/4995057ad4c938d6751c329632eb44130ac1567a/packages/repository-tests/src/crud/create-retrieve.suite.ts#L36-L42

I think we need to add `type: features.idType` to `User.id` definition, that should hopefully fix the problem you are observing.

**Comment by agnes512** [comment]
`AUTO_INCREMENT` can only apply to type `INT`. And `generated` along with `useDefaultIdType` are not enough for auto-generate string type attribute. We need to rework at **connector level** to allow so. 

I am working on [Postgres: allow string type attribute to be auto-generated ](https://github.com/strongloop/loopback-connector-postgresql/pull/404). That PR is proposing to use `uuid_generate_v4()` as a default function once the model is set to:
```
  @property({
    id: true, //optional
    type: 'string'
    generated: true,
    useDefaultIdType: false,
    postgresql: {
      dataType: 'uuid',
    },
  })
  id: String;
```
I've tested it locally, this solution works in `repository-tests` with postgresql.

Once it's done, we can rework the MySQL connector with a similar solution. Please share your idea there if you have any :smile:

**Comment by derdeka** [comment]
> That PR is proposing to use `uuid_generate_v4()` as a default function

What if someone wants to have a different generator function? (e.g. different UUID version or database extension, different function name)
I think the ddl generation should respect the defaultFn if present.

**Comment by agnes512** [comment]
> What if someone wants to have a different generator function? (e.g. different UUID version or database extension, different function name)

That's doable. It will be user's responsibility to provide valid extension and defaultFn.
And if the extension and defaultFn are not provided at the same time, it'd use `uuid_generate_v4()`. WDYT?

File: packages/repository/src/model.ts (hunk 1 above)
Comment (frbuceta) [reply packages/repository/src/model.ts]: Yes, it seems right to me. I just implemented it

**Comment by frbuceta** [comment]
> > TASK=test-repository-mongodb and TASK=test-repository-mysql fail in travis ci.
> > The `useDefaultIdType=false` seems not to be used by the juggler.
> 
> On the second thought, I think this is a bug in the way how our shared test suite is creating models.
> 
> The way how this test suite is supposed to work, is that the connector specifies what type is used for primary keys (number vs. string) and the test suite honors that type when defining new models.
> 
> I suspect the code building `User` model is not taking into account database-specific type and always uses `string`.
> 
> AFAICT, `User.id` property is defined here:
> 
> https://github.com/strongloop/loopback-next/blob/4995057ad4c938d6751c329632eb44130ac1567a/packages/repository-tests/src/crud/nested-model-properties.suite.ts#L91-L95
> 
> Here is the correct definition of the PK property in a different model:
> 
> https://github.com/strongloop/loopback-next/blob/4995057ad4c938d6751c329632eb44130ac1567a/packages/repository-tests/src/crud/create-retrieve.suite.ts#L36-L42
> 
> I think we need to add `type: features.idType` to `User.id` definition, that should hopefully fix the problem you are observing.

I've tried this but it doesn't work. Right now with the only one I have problems with is mongo.

File: packages/repository/src/model.ts [hunk 2]
```diff
@@ -106,6 +106,17 @@ export class ModelDefinition {
     const definition = (definitionOrType as PropertyDefinition).type
       ? (definitionOrType as PropertyDefinition)
       : {type: definitionOrType};
+
+    if (
+      definition.id === true &&
+      definition.generated === true &&
+      definition.type === 'string'
```
Comment (bajtos) [inline packages/repository/src/model.ts]: I am afraid we misunderstood each other. Here is what I meant:

```suggestion
      definition.type !== undefined
```

Either way, please add a new test to verify that `useDefaultIdType` is not set when `definition.type` was not provided (and was not inferred from TypeScript metadata either).

File: packages/repository/src/model.ts [hunk 3]
```diff
@@ -106,6 +106,17 @@ export class ModelDefinition {
...
+    ) {
+      Object.assign(definition, {
```
Comment (bajtos) [inline packages/repository/src/model.ts]: Nitpick (performance micro-optimization):

```ts
definition.useDefaultIdType = false;
```

**Comment by bajtos** [comment]
> > > TASK=test-repository-mongodb and TASK=test-repository-mysql fail in travis ci.
> > 
> > On the second thought, I think this is a bug in the way how our shared test suite is creating models.
> > I think we need to add `type: features.idType` to `User.id` definition, that should hopefully fix the problem you are observing.
>
> I've tried this but it doesn't work. Right now with the only one I have problems with is mongo.

Hmm...

@agnes512 IIRC, you were dealing with similar issues recently, while working on inclusion resolvers. Can you please help @frbuceta to find a way how to make repository tests pass against MongoDB?

File: packages/repository/src/model.ts (hunk 3 above)
Comment (frbuceta) [reply packages/repository/src/model.ts]: Take note

**Comment by frbuceta** [comment]
Thanks for your help @agnes512

File: packages/repository/src/model.ts (hunk 2 above)
Comment (frbuceta) [reply packages/repository/src/model.ts]: There is no model in the tests of the type as undefined, how should I generate a model with these features?

File: packages/repository/src/model.ts [hunk 4]
```diff
@@ -106,6 +106,15 @@ export class ModelDefinition {
     const definition = (definitionOrType as PropertyDefinition).type
       ? (definitionOrType as PropertyDefinition)
       : {type: definitionOrType};
+
+    if (
```
Comment (agnes512) [inline packages/repository/src/model.ts]: what if I have:
```ts
@property({
  id: true,
  generated: true,
  useDefaultIdType: true
})
id: string;
```
Would it overwrite the value of `useDefaultIdType` above?

Comment (agnes512) [reply packages/repository/src/model.ts]: I added a check `&&
      definition.useDefaultIdType === undefined`

File: packages/repository-tests/src/crud/relations/acceptance/multi-relations-inclusion-resolver.relation.acceptance.ts [hunk 5]
```diff
@@ -63,15 +70,22 @@ export function hasManyInclusionResolverAcceptance(
...
+        zipcode: '8200',
+        customerId: parent.id,
       });
       const order = await orderRepo.create({
         description: 'an order',
         customerId: parent.id,
       });
 
       const result = await customerRepo.find({
-        include: [{relation: 'orders'}, {relation: 'customers'}],
```
Comment (agnes512) [inline packages/repository-tests/src/crud/relations/acceptance/multi-relations-inclusion-resolver.relation.acceptance.ts]: While I was fixing the CI failure, I think the flag might somehow break  how we deal with `ObjectId` in MongoDB. Before we were able to traverse related instances with `ObjectId`.  Not sure it has something to do with the type.

File: packages/repository/src/model.ts (hunk 4 above)
Comment (frbuceta) [reply packages/repository/src/model.ts]: Good approach, I have seen that you have already changed it so I will give this revision as resolved

**Comment by bajtos** [comment]
@agnes512 @frbuceta what's the status of this pull request? I see that the CI is green and the code looks reasonable, even if not all edge cases may be covered by tests. Should we perhaps proceed to land the changes as they are now?
